sql/core/src/main/scala/org/apache/spark/sql/DataFrameImpl.scala

```diff
@@ -323,7 +323,7 @@ private[sql] class DataFrameImpl protected[sql](
   override def count(): Long = groupBy().count().rdd.collect().head.getLong(0)
 
   override def repartition(numPartitions: Int): DataFrame = {
-    sqlContext.applySchema(rdd.repartition(numPartitions), schema)
+    sqlContext.createDataFrame(rdd.repartition(numPartitions), schema)
```

Comment:
not part of this pr, but we should probably add a logical repartition operator so we can enable predicate pushdown for repartition.